Add Inventory card

rawagner opened the pull request:
![inventory](https://user-images.githubusercontent.com/2078045/60082541-3b454600-9734-11e9-8191-95888ddf53c7.png)

Plugins can contribute new items to Inventory tab and also new status groups via `DashboardsOverviewInventoryItem` and `DashboardsInventoryItemGroup` extension points.

Every status group is a link pointing to resource page with enabled filters (ie Pods in OK group will link to  Pods page with `'Running` and `Succeeded` status).

vojtechszocs commented:
> Plugins can contribute new items to Inventory tab

... to Cluster Inventory card within the Overview tab :smiley:

vojtechszocs commented on frontend/public/components/dashboards-page/overview-dashboard/inventory-card.tsx:
```diff
@@ -0,0 +1,117 @@
+  const allResources = [...k8sResources];
+  const pluginItems = plugins.registry.getDashboardsOverviewInventoryItems();
+  pluginItems.forEach((item, index) => {
+    allResources.push(uniqueResource(item.properties.resource,index));
```
Nit: missing space after `,` (also in code below) - please run linter :smiley:

rawagner replied:
fixed :)

vojtechszocs commented on frontend/public/components/dashboards-page/overview-dashboard/inventory-card.tsx:
```diff
@@ -0,0 +1,117 @@
+
+const getResourcesToWatch = (): FirehoseResource[] => {
+  const allResources = [...k8sResources];
+  const pluginItems = plugins.registry.getDashboardsOverviewInventoryItems();
```
```ts
getDashboardsOverviewInventoryItems().map(item => {
  if (!item.properties.resource.kind) {
    item.properties.resource.kind = item.properties.model.kind;
  }
});
```

rawagner replied:
not needed anymore, sice we are staying with `FirehoseResource` type

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +36,31 @@ namespace ExtensionProperties {
+    id: string;
+
+    /** Loader for the corresponding React component. */
+    loader: LazyLoader<any>;
```
Based on discussion, please change the type to `React.ReactElement` - similar to `Perspective.icon`.

rawagner replied:
isnt `React.ReactNode` more appropriate here ? I could also rename `icon` to `Icon` to sign that it is React component.

rawagner replied:
changed type to `React.ComponentType`

rawagner replied:
Using `React.ReactElement` same as in `Perspective.icon`

rawagner commented on frontend/packages/console-demo-plugin/src/plugin.tsx:
```diff
@@ -167,6 +172,28 @@ const plugin: Plugin<ConsumedExtensions> = [
+    properties: {
+      id: 'demo-inventory-group',
+      loader: () =>
+        import('./dashboards/inventory' /* webpackChunkName: "demo-plugin-inventory" */).then(
```
fixed

rawagner commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +36,31 @@ namespace ExtensionProperties {
   }
+
+  export interface DashboardsOverviewInventoryItem {
+    /** Resource which will be fetched and groupped by `filter` function. */
```
fixed

rawagner commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +36,31 @@ namespace ExtensionProperties {
+
+  export interface DashboardsOverviewInventoryItem {
+    /** Resource which will be fetched and groupped by `filter` function. */
+    resource: FirehoseResource;
```
`RequiredExceptFor` definition isn't correct as we are using different `Diff` definition than expected. Sticking to `FirehoseResource` for now

vojtechszocs replied:
OK, we can improve this later on.

rawagner commented on frontend/public/components/dashboard/inventory-card/inventory-item.tsx:
```diff
@@ -0,0 +1,128 @@
+import { LoadingInline, AsyncComponent } from '../../utils';
+import { K8sResourceKind, K8sKind } from '../../../module/k8s';
+
+export enum InventoryGroup {
```
fixed

rawagner commented on frontend/public/components/dashboard/inventory-card/inventory-item.tsx:
```diff
@@ -0,0 +1,128 @@
+
+const getPluginStatusGroups = () => {
+  const pluginGroups = {};
+  plugins.registry.getDashboardsInventoryItemGroups().forEach(status => {
```
forgot to change it :) Its `group` now.

rawagner commented on frontend/public/components/dashboard/inventory-card/inventory-item.tsx:
```diff
@@ -0,0 +1,128 @@
+  NOT_MAPPED = 'NOT_MAPPED',
+}
+
+const getPluginStatusGroups = () => {
```
fixed

rawagner commented on frontend/public/components/dashboard/inventory-card/inventory-item.tsx:
```diff
@@ -0,0 +1,128 @@
+
+const Status: React.FC<StatusProps> = React.memo(({ groupID, count, statusIDs, kind, namespace, filterType}) => {
+  const statusItems = encodeURIComponent(statusIDs.join(','));
+  namespace = namespace ? `ns/${namespace}` : 'all-namespaces';
```
New `const namespacePath` is used

spadgett commented:
/assign @vojtechszocs

spadgett commented:
/approve

Approving pending lgtm from @vojtechszocs

spadgett commented:
/ok-to-test

vojtechszocs commented on frontend/packages/console-demo-plugin/src/dashboards/inventory.tsx:
```diff
@@ -0,0 +1,20 @@
+  },
+});
+
+export const DemoGroup: React.FC<{}> = () => (
```
Nit: `DemoGroupIcon`.

rawagner replied:
renamed to `GemoGroupIcon`

vojtechszocs commented on frontend/packages/console-demo-plugin/src/plugin.tsx:
```diff
@@ -167,6 +172,25 @@ const plugin: Plugin<ConsumedExtensions> = [
+  {
+    type: 'Dashboards/Inventory/Item/Group',
+    properties: {
+      id: 'demo-inventory-group',
```
This doesn't match `demo-inventory-status` above in `getRouteStatusGroups` (?)

rawagner replied:
oops, sorry about that. It got out-of-sync while doing previous changes

rawagner replied:
fixed now

vojtechszocs commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -31,6 +35,31 @@ namespace ExtensionProperties {
+    useAbbr?: boolean;
+
+    /** Function which will divide various statuses to groups. */
+    filter: StatusGroupFilter;
```
Nit: this is technically a mapper, not a filter, but we can adapt it later.

rawagner replied:
right, it is mapper. I replaced all `filter` references with `mapper`

rawagner commented:
@vojtechszocs as per our discussion I removed `plugin-stats` from `package.json`. Can you take a look ?

rawagner commented:
> @vojtechszocs as per our discussion I removed `plugin-stats` from `package.json`. Can you take a look ?

added back sice https://github.com/openshift/console/pull/1871 is up

rawagner commented:
/test e2e-aws

vojtechszocs commented:
/lgtm
